source/src/protocols/cyclic_peptide/crosslinker/Lanthionine_Helper.cc

```diff
@@ -0,0 +1,311 @@
+bool
+Lanthionine_Helper::filter_by_sidechain_distance_symmetric(
+	core::pose::Pose const & pose,
+	core::select::residue_selector::ResidueSubset const & selection,
+	core::Real const & filter_multiplier
+) const {
+	//This will not compile until this is implemented.
+	return filter_by_constraints_energy( pose, selection, false, false, filter_multiplier );
+}
+
+/// @brief Determine whether the sidechain-crosslinker system has too high a constraints score.
+/// @details Returns TRUE for failure (too high a constraints score) and FALSE for success.
+/// @note Higher values of the filter multiplier make it more permissive.
+bool
+Lanthionine_Helper::filter_by_constraints_energy_asymmetric(
+	core::pose::Pose const & pose,
+	core::select::residue_selector::ResidueSubset const & selection,
+	core::Real const & filter_multiplier
+) const {
+	//The default is to call the base class filter_by_constraints_energy() function:
+	return filter_by_constraints_energy( pose, selection, false, false, filter_multiplier );
+}
+
+/// @brief Determine whether the sidechain-crosslinker system has too high a constraints
+/// score.  This version is for symmetric poses.
+/// @details Returns TRUE for failure (too high a constraints score) and FALSE for success.
+/// @note Higher values of the filter multiplier make it more permissive.
+bool
+Lanthionine_Helper::filter_by_constraints_energy_symmetric(
+	core::pose::Pose const & pose,
+	core::select::residue_selector::ResidueSubset const & selection,
+	bool const linker_was_added,
+	core::Real const & filter_multiplier
+) const {
+	//The default is to call the base class filter_by_constraints_energy() function:
+	return filter_by_constraints_energy( pose, selection, true, linker_was_added, filter_multiplier );
+}
+
+/********************************************
+PRIVATE FUNCTIONS
+*********************************************/
+///@brief Given a selection of exactly two residues and a pose, return the two lanthionine
+/// residues.
+void
+Lanthionine_Helper::get_lanthionine_residues_from_selection(
+	core::select::residue_selector::ResidueSubset const & selection,
+	core::pose::Pose const & pose,
+	core::Size & dalares,
+	core::Size & cysres
+) const {
+	//This will not compile until this is implemented.
+	std::string const errmsg( "Error in Lanthionine_Helper::get_lanthionine_residues_from_selection(): " );
+	dalares = 0;
+	cysres = 0;
+
+	debug_assert( selection.size() == pose.total_residue() );
+	debug_assert( selection.size() >= 2 );
+
+	for ( core::Size i(1), imax(selection.size()); i<=imax; ++i ) {
+		if ( selection[i] ) {
+			if ( dalares == 0 ) {
+				dalares = i;
+			} else if ( cysres == 0 ) {
+				cysres = i;
+				if ( !pose.residue_type( cysres ).is_sidechain_thiol() ) { cysres=dalares; dalares=i;}
+			} else {
+				utility_exit_with_message( errmsg + "More than two residues were selected!  A lanthionine connection is specified by exactly two residues." );
+			}
+		}
+	}
```

Comment:
The logic here is a bit confused with the multiple swapping opportunities. I would suggest tightening it up by using the get_residues_from_subset() function from core/select/util.hh. This should give you a 2-length vector, and you can then figure out which one is the cys and which one is the dala (only assigning them once.)